Scraping ... 

In [1]:
import pandas as pd

Datos obtenidos de Argenprop

In [2]:
df = pd.read_csv("datos_argenprop_raw.csv",sep=';',encoding="utf-8") 
df.head()

precio  \
0  USD 155.000 + $120.000 expensas   
1                      USD 149.000   
2  USD 550.000 + $750.000 expensas   
3  USD 395.000 + $450.000 expensas   
4                      USD 295.000   

                                              barrio  \
0  Departamento en Venta en Colegiales, Capital F...   
1  Departamento en Venta en Villa Ortuzar, Capita...   
2  Departamento en Venta en Barrio Norte, Capital...   
3  Departamento en Venta en Barrio Norte, Capital...   
4  Departamento en Venta en Villa Urquiza, Capita...   

                             caracteristicas  \
0     ['45 m² cubie.', '1 dorm.', '15 años']   
1  ['49,55 m² cubie.', '1 dorm.', '24 años']   
2    ['162 m² cubie.', '3 dorm.', '70 años']   
3    ['164 m² cubie.', '3 dorm.', '40 años']   
4      ['105 m² cubie.', '2 dorm.', '1 año']   

                                       descripcion_1  \
0                     LOFT en colegiales CON COCHERA   
1    VENTA 2 AMBIENTES VILLA ORTUZAR PARRILLA BALCON   
2  DEPARTAMENTO 3 DOR.C/DEP - BARRIO NORTE, CAPIT...   
3  Departamento 4 AMB, con Cochera Fija y Baulera...   
4    Departamento 3 ambientes a estrenar con cochera   

                                       descripcion_2  
0  ¡Loft de ensueño en la mejor zona de colegiale...  
1  Corredor responsable: graciela menazzi cmcpdjl...  
2  Pisos de calden - vigilancia 24hs. - Baño, toi...  
3  Oportunidad departamento 4 amb, sobre av. Sant...  
4  Corredor responsable: alan carlos bosque csi 6...

# Limpieza

In [3]:
len(df)

22416

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22416 entries, 0 to 22415
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   precio           22416 non-null  object
 1   barrio           22416 non-null  object
 2   caracteristicas  22416 non-null  object
 3   descripcion_1    22416 non-null  object
 4   descripcion_2    22416 non-null  object
dtypes: object(5)
memory usage: 875.8+ KB


No hay duplicados

In [5]:
df.duplicated().sum()

0

## Precio

In [6]:
# Crea una nueva columna solo con el valor USD
df['precio_usd'] = df['precio'].str.extract(r'USD\s*([\d\.]+)', expand=False)
df.head()

precio  \
0  USD 155.000 + $120.000 expensas   
1                      USD 149.000   
2  USD 550.000 + $750.000 expensas   
3  USD 395.000 + $450.000 expensas   
4                      USD 295.000   

                                              barrio  \
0  Departamento en Venta en Colegiales, Capital F...   
1  Departamento en Venta en Villa Ortuzar, Capita...   
2  Departamento en Venta en Barrio Norte, Capital...   
3  Departamento en Venta en Barrio Norte, Capital...   
4  Departamento en Venta en Villa Urquiza, Capita...   

                             caracteristicas  \
0     ['45 m² cubie.', '1 dorm.', '15 años']   
1  ['49,55 m² cubie.', '1 dorm.', '24 años']   
2    ['162 m² cubie.', '3 dorm.', '70 años']   
3    ['164 m² cubie.', '3 dorm.', '40 años']   
4      ['105 m² cubie.', '2 dorm.', '1 año']   

                                       descripcion_1  \
0                     LOFT en colegiales CON COCHERA   
1    VENTA 2 AMBIENTES VILLA ORTUZAR PARRILLA BALCON   
2  DEPARTAMENTO 3 DOR.C/DEP - BARRIO NORTE, CAPIT...   
3  Departamento 4 AMB, con Cochera Fija y Baulera...   
4    Departamento 3 ambientes a estrenar con cochera   

                                       descripcion_2 precio_usd  
0  ¡Loft de ensueño en la mejor zona de colegiale...    155.000  
1  Corredor responsable: graciela menazzi cmcpdjl...    149.000  
2  Pisos de calden - vigilancia 24hs. - Baño, toi...    550.000  
3  Oportunidad departamento 4 amb, sobre av. Sant...    395.000  
4  Corredor responsable: alan carlos bosque csi 6...    295.000

Hay valores en precio con Consultar precio. se descartan

In [7]:
df[df['precio_usd'].isna()].head()

precio  \
327                      Consultar precio   
364   Consultar precio + $20.000 expensas   
391   Consultar precio + $30.000 expensas   
393                      Consultar precio   
473  Consultar precio + $312.000 expensas   

                                                barrio  \
327   Departamento en Venta en Flores, Capital Federal   
364  Departamento en Venta en Belgrano, Capital Fed...   
391  Departamento en Venta en Liniers, Capital Federal   
393  Departamento en Venta en Villa del Parque, Cap...   
473    Departamento en Venta en Palermo Nuevo, Palermo   

                                caracteristicas  \
327    ['36 m² cubie.', 'A Estrenar', '1 baño']   
364  ['117 m² cubie.', '3 dorm.', 'A Estrenar']   
391      ['42 m² cubie.', '1 dorm.', '40 años']   
393      ['52 m² cubie.', '2 dorm.', '10 años']   
473   ['70,49 m² cubie.', '2 dorm.', '19 años']   

                                         descripcion_1  \
327            Monoambientes divisibles con 2 balcones   
364  Departamento en Venta ubicado en Belgrano, Cap...   
391                  Departamento 2 AMB. Venta LINIERS   
393     VENTA HERMOSO SEMIPISO 3 AMB. VILLA DEL PARQUE   
473    VENTA DEPARTAMENTO 3 AMBIENTES EN PALERMO NUEVO   

                                         descripcion_2 precio_usd  
327  Semipisos monoambientes divisible mudate ya 40...        NaN  
364  Se vende: excelente departamento sobre quesada...        NaN  
391  Departamento 2 amb. Frente en liniers. Primer ...        NaN  
393  Hermoso semipiso de 3 ambientes en edificio de...        NaN  
473  Corredor responsable: graciela menazzi cmcpdjl...        NaN

In [8]:
len(df[df['precio_usd'].isna()])

53

In [9]:
df.dropna(subset=['precio_usd'], inplace=True)
df.reset_index(drop=True, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22363 entries, 0 to 22362
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   precio           22363 non-null  object
 1   barrio           22363 non-null  object
 2   caracteristicas  22363 non-null  object
 3   descripcion_1    22363 non-null  object
 4   descripcion_2    22363 non-null  object
 5   precio_usd       22363 non-null  object
dtypes: object(6)
memory usage: 1.0+ MB


In [10]:
# valor como entero
df['precio_usd'] = df['precio_usd'].str.replace('.', '', regex=False).astype(int)

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22363 entries, 0 to 22362
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   precio           22363 non-null  object
 1   barrio           22363 non-null  object
 2   caracteristicas  22363 non-null  object
 3   descripcion_1    22363 non-null  object
 4   descripcion_2    22363 non-null  object
 5   precio_usd       22363 non-null  int32 
dtypes: int32(1), object(5)
memory usage: 961.0+ KB


In [12]:
df[['precio', 'precio_usd']].head()

precio  precio_usd
0  USD 155.000 + $120.000 expensas      155000
1                      USD 149.000      149000
2  USD 550.000 + $750.000 expensas      550000
3  USD 395.000 + $450.000 expensas      395000
4                      USD 295.000      295000

## Barrio

La descripcion del barrio en las publicaciones no es consistente, puede aparecer luego de la coma: Capital Federal
o Centro o un barrio como Palermo. Además tiene barrios como subbarrios en el primer termino. 

Se define funcion para manejar estas inconsistencias 

In [13]:
df['barrio']

0        Departamento en Venta en Colegiales, Capital F...
1        Departamento en Venta en Villa Ortuzar, Capita...
2        Departamento en Venta en Barrio Norte, Capital...
3        Departamento en Venta en Barrio Norte, Capital...
4        Departamento en Venta en Villa Urquiza, Capita...
                               ...                        
22358    Departamento en Venta en Villa del Parque, Cap...
22359    Departamento en Venta en Villa Crespo, Capital...
22360    Departamento en Venta en Barracas, Capital Fed...
22361    Departamento en Venta en Parque Avellaneda, Fl...
22362    Departamento en Venta en Villa Crespo, Capital...
Name: barrio, Length: 22363, dtype: object

In [14]:
def limpiar_barrio(texto):
    # Busca la última ocurrencia de " en "
    idx = texto.lower().rfind(" en ")
    if idx != -1:
        resto = texto[idx + 4:]  # Salta " en "
        partes = [p.strip() for p in resto.split(",")]
        if len(partes) == 2:
            # Caso 1: Si el segundo miembro es "Capital Federal", elegir el primero
            if partes[1].lower() == "capital federal":
                return partes[0]
            # Caso 2: Si el segundo miembro es "Centro", elegir el primero
            elif partes[1].lower() == "centro":
                return partes[0]
            # Caso 3: Si el segundo miembro es "Palermo", elegir el segundo
            elif partes[1].lower() == "palermo":
                return partes[1]
            # Caso 4: Si el segundo miembro es "Barrio Norte", elegir el segundo
            elif partes[1].lower() == "barrio norte":
                return partes[1]
            # Caso 5: Si el segundo miembro es "Floresta" y el primero es "Monte Castro", elegir el primero
            elif partes[1].lower() == "floresta" and partes[0].lower() == "monte castro":
                return partes[0]
            # Caso 6: Si el segundo miembro es "Floresta" y el primero es "Parque Avellaneda", elegir el primero
            elif partes[1].lower() == "floresta" and partes[0].lower() == "parque avellaneda":
                return partes[0]
            # Caso 7: Si el segundo miembro es "Villa Urquiza" y el primero es "Parque Chas", elegir el primero
            elif partes[1].lower() == "villa urquiza" and partes[0].lower() == "parque chas":
                return partes[0]
            # Caso por defecto: elegir el segundo miembro
            else:
                return partes[1]
        else:
            return partes[0]  # en caso que haya solo un miembro elige el primero
    return texto

In [15]:
df['barrio_caba'] = df['barrio'].apply(limpiar_barrio)
df[['barrio', 'barrio_caba']]

barrio        barrio_caba
0      Departamento en Venta en Colegiales, Capital F...         Colegiales
1      Departamento en Venta en Villa Ortuzar, Capita...      Villa Ortuzar
2      Departamento en Venta en Barrio Norte, Capital...       Barrio Norte
3      Departamento en Venta en Barrio Norte, Capital...       Barrio Norte
4      Departamento en Venta en Villa Urquiza, Capita...      Villa Urquiza
...                                                  ...                ...
22358  Departamento en Venta en Villa del Parque, Cap...   Villa del Parque
22359  Departamento en Venta en Villa Crespo, Capital...       Villa Crespo
22360  Departamento en Venta en Barracas, Capital Fed...           Barracas
22361  Departamento en Venta en Parque Avellaneda, Fl...  Parque Avellaneda
22362  Departamento en Venta en Villa Crespo, Capital...       Villa Crespo

[22363 rows x 2 columns]

## Características

In [16]:
import re

def extraer_caracteristicas(texto):
    m2 = None
    antiguedad = None
    dorm = None

    if not isinstance(texto, str):
        return {"m2_cub": m2, "antig": antiguedad, "dorm": dorm}

    # m2 cubiertos
    m2_match = re.search(r'(\d+[\.,]?\d*)\s*m[²2]?\s*cub', texto.replace(" ", ""), re.IGNORECASE)
    if m2_match:
        m2 = m2_match.group(1).replace(',', '.')

    # antigüedad
    if re.search(r"a estrenar", texto, re.IGNORECASE):
        antiguedad = 0
    else:
        antiguedad_match = re.search(r'(\d+)\s*años?', texto, re.IGNORECASE)
        if antiguedad_match:
            antiguedad = int(antiguedad_match.group(1))

    # dormitorios
    dorm_match = re.search(r'(\d+)\s*dorm(itorios)?', texto, re.IGNORECASE)
    if dorm_match:
        dorm = int(dorm_match.group(1))
    elif re.search(r'monoam', texto, re.IGNORECASE):
        dorm = 1

    return {"m2_cub": m2, "antig": antiguedad, "dorm": dorm}

Extraer características del departamento

In [17]:
df['caracteristicas'] = df['caracteristicas'].astype(str)  # por si hay valores no string
# Aplicar la función extraer_caracteristicas a cada fila
caracteristicas = df['caracteristicas'].apply(extraer_caracteristicas)
# Convertir el resultado en un DataFrame
caracteristicas_df = pd.DataFrame(caracteristicas.tolist())
# Agregar las nuevas columnas al DataFrame original
df['m2_cub'] = caracteristicas_df['m2_cub']
df['antig'] = caracteristicas_df['antig']
df['dorm'] = caracteristicas_df['dorm']
df.head()

precio  \
0  USD 155.000 + $120.000 expensas   
1                      USD 149.000   
2  USD 550.000 + $750.000 expensas   
3  USD 395.000 + $450.000 expensas   
4                      USD 295.000   

                                              barrio  \
0  Departamento en Venta en Colegiales, Capital F...   
1  Departamento en Venta en Villa Ortuzar, Capita...   
2  Departamento en Venta en Barrio Norte, Capital...   
3  Departamento en Venta en Barrio Norte, Capital...   
4  Departamento en Venta en Villa Urquiza, Capita...   

                             caracteristicas  \
0     ['45 m² cubie.', '1 dorm.', '15 años']   
1  ['49,55 m² cubie.', '1 dorm.', '24 años']   
2    ['162 m² cubie.', '3 dorm.', '70 años']   
3    ['164 m² cubie.', '3 dorm.', '40 años']   
4      ['105 m² cubie.', '2 dorm.', '1 año']   

                                       descripcion_1  \
0                     LOFT en colegiales CON COCHERA   
1    VENTA 2 AMBIENTES VILLA ORTUZAR PARRILLA BALCON   
2  DEPARTAMENTO 3 DOR.C/DEP - BARRIO NORTE, CAPIT...   
3  Departamento 4 AMB, con Cochera Fija y Baulera...   
4    Departamento 3 ambientes a estrenar con cochera   

                                       descripcion_2  precio_usd  \
0  ¡Loft de ensueño en la mejor zona de colegiale...      155000   
1  Corredor responsable: graciela menazzi cmcpdjl...      149000   
2  Pisos de calden - vigilancia 24hs. - Baño, toi...      550000   
3  Oportunidad departamento 4 amb, sobre av. Sant...      395000   
4  Corredor responsable: alan carlos bosque csi 6...      295000   

     barrio_caba m2_cub  antig  dorm  
0     Colegiales     45   15.0   1.0  
1  Villa Ortuzar  49.55   24.0   1.0  
2   Barrio Norte    162   70.0   3.0  
3   Barrio Norte    164   40.0   3.0  
4  Villa Urquiza    105    1.0   2.0

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22363 entries, 0 to 22362
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   precio           22363 non-null  object 
 1   barrio           22363 non-null  object 
 2   caracteristicas  22363 non-null  object 
 3   descripcion_1    22363 non-null  object 
 4   descripcion_2    22363 non-null  object 
 5   precio_usd       22363 non-null  int32  
 6   barrio_caba      22363 non-null  object 
 7   m2_cub           21765 non-null  object 
 8   antig            18877 non-null  float64
 9   dorm             19151 non-null  float64
dtypes: float64(2), int32(1), object(7)
memory usage: 1.6+ MB


In [19]:
# Filas donde alguna de las columnas extraídas es nula
df_irregulares = df[df['m2_cub'].isna() | df['antig'].isna()]
# Mostrar las primeras filas con sus características
df_irregulares[['caracteristicas', 'm2_cub', 'antig']].head()

caracteristicas m2_cub  antig
17    ['47 m² cubie.', '1 dorm.', '1 baño']     47    NaN
20  ['119 m² cubie.', '3 dorm.', '3 baños']    119    NaN
21    ['28 m² cubie.', '1 baño', 'Monoam.']     28    NaN
22   ['73 m² cubie.', '2 dorm.', '2 baños']     73    NaN
23    ['28 m² cubie.', '1 baño', 'Monoam.']     28    NaN

Metros cubiertos

filas con nulos

In [20]:
print(len(df[df['m2_cub'].isna()]))
df[df['m2_cub'].isna()].head()

598


precio  \
88     USD 39.800 + $28.000 expensas   
93   USD 118.000 + $160.000 expensas   
94                        USD 47.000   
131  USD 110.000 + $111.000 expensas   
132                      USD 310.000   

                                                barrio  \
88          Departamento en Venta en Monserrat, Centro   
93          Departamento en Venta en Monserrat, Centro   
94          Departamento en Venta en Monserrat, Centro   
131  Departamento en Venta en Palermo, Capital Federal   
132  Departamento en Venta en Palermo Hollywood, Pa...   

                          caracteristicas  \
88   ['1 dorm.', '1 baño', '2 ambientes']   
93   ['2 dorm.', '1 baño', '4 ambientes']   
94                                     []   
131       ['7 años', '1 baño', 'Monoam.']   
132   ['1 dorm.', 'A Estrenar', '1 baño']   

                                         descripcion_1  \
88            DEPTO 2 AMBIENTES CON TERRAZA Y LAVAERO.   
93              DUPLEX 4 AMBIENTES EN MUY BUEN ESTADO.   
94   DEPARTAMENTO DE 2 AMBIENTES EN PLANTA BAJA CON...   
131  1 amb. divis.bcon. c/fte. Apto Prof. y/o vvda....   
132                   Departamento - Palermo Hollywood   

                                         descripcion_2  precio_usd  \
88   Departamento en planta baja 30 m2 total 50 m2 ...       39800   
93   Departamento en duplex - 2 dormitorios con pla...      118000   
94   Departamento tipo casa en planta baja con pati...       47000   
131  Caracteristicas grales.: 1 ½ amb. Divisible, b...      110000   
132  Departamento 2 ambientes en venta palermo holl...      310000   

    barrio_caba m2_cub  antig  dorm  
88    Monserrat   None    NaN   1.0  
93    Monserrat   None    NaN   2.0  
94    Monserrat   None    NaN   NaN  
131     Palermo   None    7.0   1.0  
132     Palermo   None    0.0   1.0

Reemplazo nulos en m2_cub con valores m2 si aparecen en descripcion_1

In [21]:
# Extraer metros cuadrados de la descripción2 cuando m2_cub es None
def extract_m2(descripcion):
    if pd.isna(descripcion):
        return None

    # Busca patrones como "50 m2", "50m2", "50M2", "50 metros cuadrados", "50 m²"
    pattern = r'(\d+(?:[\.,]\d+)?)\s*(m2|m²|metros cuadrados)'
    match = re.search(pattern, descripcion.lower())
    if match:
        return float(match.group(1).replace(',', '.'))
    return None

# Aplicar la función a las filas donde m2_cub es None
mask = df['m2_cub'].isna()
df.loc[mask, 'm2_cub'] = df.loc[mask, 'descripcion_1'].apply(extract_m2)


Reemplazo nulos en m2_cub con valores m2 si aparecen en descripcion_2

In [22]:
# Aplicar la función a las filas donde m2_cub es None
mask = df['m2_cub'].isna()
df.loc[mask, 'm2_cub'] = df.loc[mask, 'descripcion_2'].apply(extract_m2)

In [23]:
# filas con valor nulo
print(len(df[df['m2_cub'].isna()]))
df[df['m2_cub'].isna()].head()

462


precio  \
93   USD 118.000 + $160.000 expensas   
94                        USD 47.000   
131  USD 110.000 + $111.000 expensas   
132                      USD 310.000   
149    USD 58.000 + $61.000 expensas   

                                                barrio  \
93          Departamento en Venta en Monserrat, Centro   
94          Departamento en Venta en Monserrat, Centro   
131  Departamento en Venta en Palermo, Capital Federal   
132  Departamento en Venta en Palermo Hollywood, Pa...   
149  Departamento en Venta en Belgrano, Capital Fed...   

                          caracteristicas  \
93   ['2 dorm.', '1 baño', '4 ambientes']   
94                                     []   
131       ['7 años', '1 baño', 'Monoam.']   
132   ['1 dorm.', 'A Estrenar', '1 baño']   
149      ['1 dorm.', '30 años', '1 baño']   

                                         descripcion_1  \
93              DUPLEX 4 AMBIENTES EN MUY BUEN ESTADO.   
94   DEPARTAMENTO DE 2 AMBIENTES EN PLANTA BAJA CON...   
131  1 amb. divis.bcon. c/fte. Apto Prof. y/o vvda....   
132                   Departamento - Palermo Hollywood   
149                                 Belgrano / Coghlan   

                                         descripcion_2  precio_usd  \
93   Departamento en duplex - 2 dormitorios con pla...      118000   
94   Departamento tipo casa en planta baja con pati...       47000   
131  Caracteristicas grales.: 1 ½ amb. Divisible, b...      110000   
132  Departamento 2 ambientes en venta palermo holl...      310000   
149  Excelente monoambiente belgrano / coghlan manu...       58000   

    barrio_caba m2_cub  antig  dorm  
93    Monserrat    NaN    NaN   2.0  
94    Monserrat    NaN    NaN   NaN  
131     Palermo    NaN    7.0   1.0  
132     Palermo    NaN    0.0   1.0  
149    Belgrano    NaN   30.0   1.0

No se puede extraer m2 de la descripcion porque no está publicado

ejemplo en fila 93

In [24]:
df.iloc[93]['descripcion_2']

'Departamento en duplex - 2 dormitorios con placares - living y comedor cocina independiente amplio lavadero baño completo con antebaño calefaccion central super luminoso vista panoramica estratégica ubicación excelente estado (la unidad se caracteriza por tener el uso exclusivo del espacio aereo por reglamento ideal para hacer jardin de invierno deck, etc.) Apto profesional - posible tomar 1'

In [25]:
df.iloc[93]['descripcion_1']

'DUPLEX 4 AMBIENTES EN MUY BUEN ESTADO.'

si quedaron nulos es porque no habia m2 en ninguna descripcion

se eliminan filas con nulos en m2_cub

In [26]:
df.dropna(subset=['m2_cub'], inplace=True)
df.reset_index(drop=True, inplace=True)
df["m2_cub"] = df["m2_cub"].astype(float)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21901 entries, 0 to 21900
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   precio           21901 non-null  object 
 1   barrio           21901 non-null  object 
 2   caracteristicas  21901 non-null  object 
 3   descripcion_1    21901 non-null  object 
 4   descripcion_2    21901 non-null  object 
 5   precio_usd       21901 non-null  int32  
 6   barrio_caba      21901 non-null  object 
 7   m2_cub           21901 non-null  float64
 8   antig            18513 non-null  float64
 9   dorm             18712 non-null  float64
dtypes: float64(3), int32(1), object(6)
memory usage: 1.6+ MB


In [27]:
df.describe()['m2_cub']

count    21901.000000
mean        63.842534
std         43.260804
min          1.000000
25%         37.000000
50%         51.000000
75%         75.000000
max        621.000000
Name: m2_cub, dtype: float64

In [28]:
len(df[df['m2_cub']<15])

34

Publicacion validas a partir de 15m2

In [29]:
df = df[df['m2_cub']>15]
df.reset_index(drop=True, inplace=True)

In [30]:
df.sort_values(by='m2_cub').head()

precio  \
10565   USD 45.000 + $50.000 expensas   
21441   USD 36.000 + $23.000 expensas   
6544    USD 40.000 + $68.000 expensas   
13004  USD 29.000 + $120.000 expensas   
9167                       USD 32.000   

                                                  barrio  \
10565            Departamento en Venta en Retiro, Centro   
21441  Departamento en Venta en Villa Crespo, Capital...   
6544   Departamento en Venta en Balvanera, Capital Fe...   
13004  Departamento en Venta en Pompeya, Capital Federal   
9167   Departamento en Venta en Parque Patricios, Cap...   

                                 caracteristicas  \
10565  ['15,58 m² cubie.', '1 dorm.', '58 años']   
21441      ['16 m² cubie.', '1 baño', 'Monoam.']   
6544      ['16 m² cubie.', '1 dorm.', '38 años']   
13004     ['16 m² cubie.', '1 dorm.', '10 años']   
9167    ['16 m² cubie.', 'A Estrenar', '1 baño']   

                                           descripcion_1  \
10565  DPTO IDEAL INVERSORES -TEMPORARIO -S. NICOLAS ...   
21441  Monoambiente en Villa Crespo con balcón France...   
6544   Venta Departamento 2 ambientes en Balvanera Id...   
13004  Oportunidad! dpto monoambiente a 2 de Av. La P...   
9167              Monoambiente en pleno Parque Patricios   

                                           descripcion_2  precio_usd  \
10565  Corredor responsable: hernan perrone cucicba 2...       45000   
21441  En una de las mejores zonas de villa crespo (v...       36000   
6544   Departamento de dos ambientes reciclado a nuev...       40000   
13004  Corredor responsable: francisco d\'atri cpmcal...       29000   
9167   Corredor responsable: rene vera cucicba 6778 /...       32000   

            barrio_caba  m2_cub  antig  dorm  
10565            Retiro   15.58   58.0   1.0  
21441      Villa Crespo   16.00    NaN   1.0  
6544          Balvanera   16.00   38.0   1.0  
13004           Pompeya   16.00   10.0   1.0  
9167   Parque Patricios   16.00    0.0   NaN

## Antiguedad

In [31]:
print(len(df[df['antig'].isna()]))
df[df['antig'].isna()].head()

3380


precio  \
17                     USD 75.705   
20                    USD 483.097   
21                    USD 106.105   
22                    USD 275.502   
23  USD 52.000 + $90.000 expensas   

                                               barrio  \
17  Departamento en Venta en Palermo, Capital Federal   
20  Departamento en Venta en Palermo, Capital Federal   
21  Departamento en Venta en Palermo, Capital Federal   
22  Departamento en Venta en Palermo, Capital Federal   
23            Departamento en Venta en Retiro, Centro   

                            caracteristicas  \
17    ['47 m² cubie.', '1 dorm.', '1 baño']   
20  ['119 m² cubie.', '3 dorm.', '3 baños']   
21    ['28 m² cubie.', '1 baño', 'Monoam.']   
22   ['73 m² cubie.', '2 dorm.', '2 baños']   
23    ['28 m² cubie.', '1 baño', 'Monoam.']   

                                        descripcion_1  \
17  Departamentos en Venta - A Estrenar - Palermo ...   
20                             Departamento - Palermo   
21                             Departamento - Palermo   
22                             Departamento - Palermo   
23  Departamento - Venta - Argentina, Capital Fede...   

                                        descripcion_2  precio_usd barrio_caba  \
17  Edificio doble frente - cordoba 4042 y estado ...       75705     Palermo   
20  Hermoso 4 ambientes al frente, el mismo cuenta...      483097     Palermo   
21  1 ambiente divisible ubicado al frente, el mis...      106105     Palermo   
22  Hermoso 3 ambientes al contrafrente, el mismo ...      275502     Palermo   
23  Luminoso monoambiente en retiro - apto credito...       52000      Retiro   

    m2_cub  antig  dorm  
17    47.0    NaN   1.0  
20   119.0    NaN   3.0  
21    28.0    NaN   1.0  
22    73.0    NaN   2.0  
23    28.0    NaN   1.0

In [32]:
# Extraer antigüedad de las descripciones cuando antig es None
def extract_antig(descripcion):
    if pd.isna(descripcion):
        return None
    desc = descripcion.lower()
    # Primero verificar si es a estrenar
    if "a estrenar" in desc:
        return 0
    # Buscar número de años, permitiendo punto o coma después
    match = re.search(r'(\d+)\s*año[s]?[.,]?', desc)
    if match:
        return int(match.group(1))
    return None

# Aplicar la función a las filas donde antig es None, primero en descripcion_1 y luego en descripcion_2
mask = df['antig'].isna()
df.loc[mask, 'antig'] = df.loc[mask, 'descripcion_1'].apply(extract_antig)
mask = df['antig'].isna()
df.loc[mask, 'antig'] = df.loc[mask, 'descripcion_2'].apply(extract_antig)

In [33]:
print(len(df[df['antig'].isna()]))
df[df['antig'].isna()].head()

3054


precio  \
20                    USD 483.097   
21                    USD 106.105   
22                    USD 275.502   
23  USD 52.000 + $90.000 expensas   
24                    USD 179.843   

                                               barrio  \
20  Departamento en Venta en Palermo, Capital Federal   
21  Departamento en Venta en Palermo, Capital Federal   
22  Departamento en Venta en Palermo, Capital Federal   
23            Departamento en Venta en Retiro, Centro   
24  Departamento en Venta en Palermo, Capital Federal   

                            caracteristicas  \
20  ['119 m² cubie.', '3 dorm.', '3 baños']   
21    ['28 m² cubie.', '1 baño', 'Monoam.']   
22   ['73 m² cubie.', '2 dorm.', '2 baños']   
23    ['28 m² cubie.', '1 baño', 'Monoam.']   
24    ['46 m² cubie.', '1 dorm.', '1 baño']   

                                        descripcion_1  \
20                             Departamento - Palermo   
21                             Departamento - Palermo   
22                             Departamento - Palermo   
23  Departamento - Venta - Argentina, Capital Fede...   
24                             Departamento - Palermo   

                                        descripcion_2  precio_usd barrio_caba  \
20  Hermoso 4 ambientes al frente, el mismo cuenta...      483097     Palermo   
21  1 ambiente divisible ubicado al frente, el mis...      106105     Palermo   
22  Hermoso 3 ambientes al contrafrente, el mismo ...      275502     Palermo   
23  Luminoso monoambiente en retiro - apto credito...       52000      Retiro   
24  Muy lindo 2 ambientes al frente, el mismo cuen...      179843     Palermo   

    m2_cub  antig  dorm  
20   119.0    NaN   3.0  
21    28.0    NaN   1.0  
22    73.0    NaN   2.0  
23    28.0    NaN   1.0  
24    46.0    NaN   1.0

3054 casos donde no aparece la antiguedad en la publicacion. 

Elimino valores nulos en antig (si no se logró obtener el dato es que no figura en caracteristicas o descripcion del inmueble) 

In [34]:
df.dropna(subset=['antig'],inplace=True)
df['antig'] = df['antig'].astype(int)
df.reset_index(drop=True, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18812 entries, 0 to 18811
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   precio           18812 non-null  object 
 1   barrio           18812 non-null  object 
 2   caracteristicas  18812 non-null  object 
 3   descripcion_1    18812 non-null  object 
 4   descripcion_2    18812 non-null  object 
 5   precio_usd       18812 non-null  int32  
 6   barrio_caba      18812 non-null  object 
 7   m2_cub           18812 non-null  float64
 8   antig            18812 non-null  int32  
 9   dorm             15667 non-null  float64
dtypes: float64(2), int32(2), object(6)
memory usage: 1.3+ MB


In [35]:
round(df.describe(),)

precio_usd   m2_cub    antig     dorm
count     18812.0  18812.0  18812.0  15667.0
mean     185872.0     64.0     70.0      2.0
std      210397.0     44.0    411.0      1.0
min       22000.0     16.0      0.0      1.0
25%       89000.0     37.0      0.0      1.0
50%      129450.0     51.0     20.0      2.0
75%      203000.0     76.0     47.0      2.0
max     4600000.0    621.0  40000.0     22.0

Valores erroneos en antiguedad en algunas publicaciones

In [36]:
df[df['antig']==40000]

precio  \
14818  USD 69.500 + $20.000 expensas   

                                                  barrio  \
14818  Departamento en Venta en Balvanera, Capital Fe...   

                                 caracteristicas  \
14818  ['50 m² cubie.', '1 dorm.', '40000 años']   

                                           descripcion_1  \
14818  Venta Departamento Balvanera Dos Ambientes *AP...   

                                           descripcion_2  precio_usd  \
14818  -Dos ambientes *apto credito* -cocina completa...       69500   

      barrio_caba  m2_cub  antig  dorm  
14818   Balvanera    50.0  40000   1.0

valor erroneo antiguedad en caracteristicas y no figura en descripcion

In [37]:
df = df[df['antig']<40000]

en caracteristicas puede aparecer antiguedad como año de construcción de la propiedad

In [38]:
df[df['antig']>1900].sort_values('antig',ascending=False).tail(5)

precio  \
5615                       USD 124.900   
6406   USD 120.000 + $170.000 expensas   
13648  USD 115.000 + $280.000 expensas   
5350                        USD 80.000   
11585  USD 135.000 + $109.000 expensas   

                                                  barrio  \
5615       Departamento en Venta en Las Cañitas, Palermo   
6406   Departamento en Venta en Urquiza R, Villa Urquiza   
13648      Departamento en Venta en Las Cañitas, Palermo   
5350   Departamento en Venta en Balvanera, Capital Fe...   
11585  Departamento en Venta en Almagro, Capital Federal   

                                   caracteristicas  \
5615      ['58 m² cubie.', '2 dorm.', '1975 años']   
6406      ['50 m² cubie.', '2 dorm.', '1975 años']   
13648     ['60 m² cubie.', '2 dorm.', '1965 años']   
5350      ['58 m² cubie.', '2 dorm.', '1965 años']   
11585  ['62,37 m² cubie.', '2 dorm.', '1945 años']   

                                           descripcion_1  \
5615   OPORTUNIDAD UNICA 3 AMB C/ BALCON . LIMITE BEL...   
6406   Venta Departamento 3 Ambientes en Villa Urquiz...   
13648       Venta Departamento 3 ambientes - Las Cañitas   
5350    Departamento a la venta en Balvanera 3 Ambientes   
11585   ¡DEPARTAMENTO 3 AMBIENTES CON BALCÓN EN ALMAGRO!   

                                           descripcion_2  precio_usd  \
5615   Departamento de tres ambientes con balcón en q...      124900   
6406   Departamento 3 ambientes – ideal inversión en ...      120000   
13648  Departamento a la venta de 3 ambientes en exce...      115000   
5350   Excelente departamento de 3 ambientes se encue...       80000   
11585  Ubicado en pringles al 200, en un tercer piso ...      135000   

         barrio_caba  m2_cub  antig  dorm  
5615         Palermo   58.00   1975   2.0  
6406   Villa Urquiza   50.00   1975   2.0  
13648        Palermo   60.00   1965   2.0  
5350       Balvanera   58.00   1965   2.0  
11585        Almagro   62.37   1945   2.0

corrijo el valor en antig

In [39]:
def corregir_antiguedad(valor):
    if valor > 1900:
        return 2025 - valor
    return valor

df['antig'] = df['antig'].apply(corregir_antiguedad)

In [40]:
df['antig'].sort_values(ascending=False).head(10)

7087     914
7183     141
10172    134
12820    125
9134     125
141      125
15186    125
4031     124
11394    119
14754    115
Name: antig, dtype: int64

publicacion con antiguedad erronea, no figura en descripcion por lo que se elimina

In [41]:
df.iloc[7087]

precio                                                   USD 120.000
barrio               Departamento en Venta en Almagro Norte, Almagro
caracteristicas              ['41 m² cubie.', '1 dorm.', '914 años']
descripcion_1      VENTA Departamento 2 amb. en Abasto, Almagro /...
descripcion_2      Para más información y propiedades similares, ...
precio_usd                                                    120000
barrio_caba                                                  Almagro
m2_cub                                                          41.0
antig                                                            914
dorm                                                             1.0
Name: 7087, dtype: object

In [42]:
df.iloc[7087]['descripcion_2']

'Para más información y propiedades similares, contactate al whatsapp . Departamento de dos ambientes, muy lindo en zona abasto, frente al shopping. Se vende amoblado. Living comedor con salida al balcón. Cocina completa integrada al living. Dormitorio. Baño amplio con box de ducha y ventilación al exterior. El edificio cuenta con amenities: piscina en la terraza y laundry. En pb hay una peatonal'

In [43]:
indice = df.iloc[7087].name
df = df.drop(indice).reset_index(drop=True)

In [44]:
df['antig'].sort_values(ascending=False).head(10)

7182     141
10171    134
141      125
9133     125
15184    125
12819    125
4031     124
11393    119
14753    115
808      115
Name: antig, dtype: int64

corrijo el valor en caractaristicas

In [45]:
def reemplazar_anio_por_antig(carac, antig):
    # Busca patrones como "19xx años" o "20xx años"
    return re.sub(r'((19|20)\d{2})\s*año[s]?', f"{int(antig)} años", carac)

df['caracteristicas'] = df.apply(lambda row: reemplazar_anio_por_antig(row['caracteristicas'], row['antig']), axis=1)

## Dormitorios

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18810 entries, 0 to 18809
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   precio           18810 non-null  object 
 1   barrio           18810 non-null  object 
 2   caracteristicas  18810 non-null  object 
 3   descripcion_1    18810 non-null  object 
 4   descripcion_2    18810 non-null  object 
 5   precio_usd       18810 non-null  int32  
 6   barrio_caba      18810 non-null  object 
 7   m2_cub           18810 non-null  float64
 8   antig            18810 non-null  int64  
 9   dorm             15665 non-null  float64
dtypes: float64(2), int32(1), int64(1), object(6)
memory usage: 1.4+ MB


In [47]:
df[df['dorm'].isna()].head()

precio  \
6   USD 110.000 + $136.000 expensas   
9    USD 110.000 + $92.000 expensas   
10  USD 105.000 + $190.000 expensas   
18   USD 59.990 + $100.000 expensas   
20                      USD 492.765   

                                               barrio  \
6   Departamento en Venta en Villa Urquiza, Capita...   
9    Departamento en Venta en Flores, Capital Federal   
10  Departamento en Venta en Palermo, Capital Federal   
18  Departamento en Venta en San Cristobal, Capita...   
20  Departamento en Venta en Palermo, Capital Federal   

                             caracteristicas  \
6      ['40 m² cubie.', '15 años', '1 baño']   
9   ['34 m² cubie.', 'A Estrenar', '1 baño']   
10     ['35 m² cubie.', '10 años', '1 baño']   
18       ['29 m² cubie.', '1 año', '1 baño']   
20               ['140 m² cubie.', 'Frente']   

                                        descripcion_1  \
6   Monoambiente 40 m.Villa Urquiza.balcón.pileta....   
9   A ESTRENAR MONOAMBIENTE CON COCHERA FIJA CONTR...   
10  Oportunidad Monoambiente, Con Amenities Premiu...   
18  VENTA DE MONO TOTALMENTE ELECTRICO FULL AMENITIES   
20                             Departamento - Palermo   

                                        descripcion_2  precio_usd  \
6   El mejor de todos! Monoambiente de 40 mt con p...      110000   
9   Av. Carabobo entre juan b. Alberdi y ramón fal...      110000   
10  Venta monoambiente divisible en edificio moder...      105000   
18  Corredor responsable: luis rodriguez de filipp...       59990   
20  Local comercial en edificio a estrenar de dobl...      492765   

      barrio_caba  m2_cub  antig  dorm  
6   Villa Urquiza    40.0     15   NaN  
9          Flores    34.0      0   NaN  
10        Palermo    35.0     10   NaN  
18  San Cristobal    29.0      1   NaN  
20        Palermo   140.0      0   NaN

In [48]:
# Extraer dormitorios de las descripciones cuando dorm es None
def extract_dorm(descripcion):
    if pd.isna(descripcion):
        return None
    desc = descripcion.lower()

    # 1. Buscar si es monoambiente o mono
    if re.search(r'monoamb|mono ambiente|monoambiente|^mono\b', desc):
        return 1
    # 2. Buscar cantidad de dormitorios (ej: 2 dorm, 3 dormitorios)
    match = re.search(r'(\d+)\s*dorm(itorio|itorios)?', desc)
    if match:
        return int(match.group(1))
    # 3. Buscar cantidad de ambientes (ej: 2 ambientes, 3 ambiente, 2 amb)
    match = re.search(r'(\d+)\s*amb(iente(s)?)?', desc)
    if match:
        return int(match.group(1))
    # 4. Si aparece solo "ambiente" o "amb" (sin número), dorm = 1
    if re.search(r'\bamb(iente)?(s)?\b', desc):
        return 1
    
    return None

# Aplicar la función a las filas donde dorm es None, primero en descripcion_1 y luego en descripcion_2
mask = df['dorm'].isna()
df.loc[mask, 'dorm'] = df.loc[mask, 'descripcion_1'].apply(extract_dorm)
mask = df['dorm'].isna()
df.loc[mask, 'dorm'] = df.loc[mask, 'descripcion_2'].apply(extract_dorm)

In [49]:
print(len(df[df['dorm'].isna()]))
df[df['dorm'].isna()].head()

116


precio                                             barrio  \
20    USD 492.765  Departamento en Venta en Palermo, Capital Federal   
185   USD 780.000            Departamento en Venta en Retiro, Centro   
933   USD 110.000       Departamento en Venta en San Nicolás, Centro   
1081  USD 175.000     Departamento en Venta en Puerto Madero, Centro   
1158  USD 210.000         Departamento en Venta en Monserrat, Centro   

                               caracteristicas  \
20                 ['140 m² cubie.', 'Frente']   
185                          ['253 m² cubie.']   
933   ['27 m² cubie.', 'A Estrenar', '1 baño']   
1081  ['32 m² cubie.', 'A Estrenar', '1 baño']   
1158  ['48 m² cubie.', 'A Estrenar', '1 baño']   

                                          descripcion_1  \
20                               Departamento - Palermo   
185                                       Arroyo al 800   
933   Ultimas unidades disponibles en Moreno 500 - P...   
1081  Departamento en Venta a Estrenar en Puerto Madero   
1158    Venta Departamento en Monserrat en construcción   

                                          descripcion_2  precio_usd  \
20    Local comercial en edificio a estrenar de dobl...      492765   
185   Edificio emblemático, patrimonial, con más de ...      780000   
933   Paseo de la cisterna un moderno edificio en el...      110000   
1081  A estrenar! Excelente departamento de lujo de ...      175000   
1158  Grand atlántida residences se encuentra en con...      210000   

        barrio_caba  m2_cub  antig  dorm  
20          Palermo   140.0      0   NaN  
185          Retiro   253.0    100   NaN  
933     San Nicolás    27.0      0   NaN  
1081  Puerto Madero    32.0      0   NaN  
1158      Monserrat    48.0      0   NaN

Elimino valores nulos en dorm (si no se logró obtener el dato es que no figura en caracteristicas o descripcion del inmueble) 

In [50]:
df.dropna(subset=['dorm'],inplace=True)
df['dorm'] = df['dorm'].astype(int)
df.reset_index(drop=True, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18694 entries, 0 to 18693
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   precio           18694 non-null  object 
 1   barrio           18694 non-null  object 
 2   caracteristicas  18694 non-null  object 
 3   descripcion_1    18694 non-null  object 
 4   descripcion_2    18694 non-null  object 
 5   precio_usd       18694 non-null  int32  
 6   barrio_caba      18694 non-null  object 
 7   m2_cub           18694 non-null  float64
 8   antig            18694 non-null  int64  
 9   dorm             18694 non-null  int32  
dtypes: float64(1), int32(2), int64(1), object(6)
memory usage: 1.3+ MB


In [51]:
df.describe()['dorm']

count    18694.000000
mean         1.731304
std          0.921150
min          0.000000
25%          1.000000
50%          1.000000
75%          2.000000
max         22.000000
Name: dorm, dtype: float64

In [52]:
df[df['dorm']==0]

precio  \
2307  USD 105.000 + $140.000 expensas   

                                                 barrio  \
2307  Departamento en Venta en Palermo, Capital Federal   

                               caracteristicas  \
2307  ['39,38 m² cubie.', '10 años', '1 baño']   

                                        descripcion_1  \
2307  VENTA MON0AMBIENTE PALERMO APTO CREDITO PISCINA   

                                          descripcion_2  precio_usd  \
2307  Corredor responsable: zmz bienes raices s.R.L....      105000   

     barrio_caba  m2_cub  antig  dorm  
2307     Palermo   39.38     10     0

In [53]:
df.loc[df['dorm']==0,'dorm'] = 1

In [54]:
df.sort_values(by='dorm')[['precio','descripcion_1','dorm']].tail(10)

precio  \
11384  USD 1.296.000 + $1.185.000 expensas   
10580                          USD 195.000   
8246       USD 420.000 + $564.000 expensas   
8330       USD 359.000 + $327.866 expensas   
14217      USD 128.000 + $194.000 expensas   
9891       USD 450.000 + $800.000 expensas   
13675      USD 247.000 + $210.000 expensas   
13436                          USD 670.000   
8374                         USD 1.800.000   
8625                           USD 850.000   

                                           descripcion_1  dorm  
11384  Venta Departamento 7 Ambientes Piso Alto al Ri...     7  
10580  Venta Local y 4 Unidades de 2 Ambientes en Pb ...     8  
8246   VENTA DPTO 450 m2- APTO PROF MONSERRAT-Tza propia     8  
8330   2 Departamentos 4 y 5 ambientes Luminosos Cong...     8  
14217        VENTA DEPTO apto profesional en Microcentro     8  
9891   EXCELENTE PISO EN VENTA APTO./PROF. AV. SANTA ...     8  
13675        Venta departamento multi familiar balvanera    10  
13436               Departamentos para renta en Recoleta    10  
8374                          ESPECTACULAR PISO 600 mts2    12  
8625   Petit Hotel de Estilo Francés en venta en Reco...    22

In [55]:
df.iloc[3688] # es error deberia ser 1 ambiente

precio                                USD 52.000 + $112.000 expensas
barrio             Departamento en Venta en Balvanera, Capital Fe...
caracteristicas             ['30 m² cubie.', 'A Estrenar', '1 baño']
descripcion_1        Departamento en Congreso de un ambiente y medio
descripcion_2      Departamento de 1 12 ambiente con entrepiso en...
precio_usd                                                     52000
barrio_caba                                                Balvanera
m2_cub                                                          30.0
antig                                                              0
dorm                                                               1
Name: 3688, dtype: object

In [56]:
df.at[3688,'dorm'] = 1

# Guardo dataset


In [57]:
df.to_csv('datos_departamentos.csv',sep=';',encoding='utf-8',index=False)